src/Microsoft.AspNet.Razor.Runtime/TagHelpers/ReadOnlyTagHelperAttributes.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
+{
+    public class ReadOnlyTagHelperAttributes<TAttribute> : IReadOnlyList<TAttribute>
+        where TAttribute : IReadOnlyTagHelperAttribute
+    {
+        protected ReadOnlyTagHelperAttributes()
+        {
+            Attributes = new List<TAttribute>();
+        }
+
+        public ReadOnlyTagHelperAttributes([NotNull] IEnumerable<TAttribute> attributes)
+        {
+            Attributes = new List<TAttribute>(attributes);
+        }
+
+        protected List<TAttribute> Attributes { get; }
+
+        public TAttribute this[int index]
+        {
+            get
+            {
+                return Attributes[index];
+            }
+        }
+
+        public TAttribute this[[NotNull] string key]
+        {
+            get
+            {
+                return Attributes.FirstOrDefault(attribute => KeyEquals(key, attribute));
```

Comment:
It may be pre-mature at this point, but I expect all of these things to need a rewrite without the linq/capturing lambda once we do some perf analysis.

You often can't really anticipate how users will get into trouble with a framework-provided collection type so it pays to take 5 sec to remove the allocation.